lib/posix/pthread.c

```diff
@@ -259,6 +273,9 @@ static void posix_thread_finalize(struct posix_thread *t, void *retval)
 	t->retval = retval;
 	k_spin_unlock(&pthread_pool_lock, key);
 
+	/* trigger recycle work */
+	(void)k_work_schedule(&posix_thread_recycle_work, K_MSEC(100));
```

Comment:
This should probably have a Kconfig associated with it. Will wait for additional reviews before a respin though.

Reply:
Why does it need a timer at all?  Can't you do the cleanup synchronously?

Reply:
Technically yes, except not in the thread itself and not if we want to avoid losing chunks of the heap or elements from the stack pool for indeterminant amount of time.

We used to have a hook in the kernel for running an arbitrary handler once a thread exited (added by @andrewboie, IIRC) but it tended to be problematic for some reason and was removed. A good indicator as to how long I've been hacking  at this 😅 

So the next be solution that I could come up (without kernel involvement) with is cleaning up pthread stacks both asynchronously (swpawned from `posix_thread_finalize()`) and on demand (in `pthread_create()`).

Auto-thread-stack cleanup (at the kernel level) would be pretty nice though, and was requested by @keith-packard as well, but that unfortunately clashes with the safety-critical requirements again, that no dynamic allocations / frees happen in core zephyr.

Definitely open to suggestions.

Reply:
yeah, to make this work you really do need a callback run at thread termination time which doesn't run on the thread stack.  If the stack was allocated from the heap, that memory won't be available until the interval expires or another call to `pthread_create`. Not great, but without some synchronous callback at thread termination, I'm not sure what the options are.

Reply:
For reference, the callback used to be stored in `fn_abort`. The commit that added it was #27995 and the commit that removed it was #29064

Reply:
I may be misremembering, but I believe the issue with `fn_abort` was when multiple conflicting APIs attempted to specify  an "abort function" in tdata (e.g. CMSIS and POSIX).

POSIX gets around this with `pthread_cleanup_push()` / `pthread_cleanup_pop()`. So for dynamically allocated thread stacks, we could push the thread stack cleanup handler. However, we shouldn't expect the universe to subscribe to the POSIX model, so that does not necessarily fix the problem of having multiple conflicting APIs trying to specify an "abort function".

Like a few things recently, it might be the case that the better place to solve that problem is in the kernel, so that we have a unified way of doing it regardless of OS abstraction / threading library.

@andyross - would it be acceptable to farm-off that work until later in it's own dedicated PR? I was [planning](https://github.com/zephyrproject-rtos/zephyr/issues/51211) on implementing `pthread_cleanup_push()` / `pthread_cleanup_pop()` anyway at some point, and this would make that significantly easier, and then pthread cleanup can also be simplified in the process.

The specific areas that I would want to be cautious of in that deferred are security-related, cc @ceolin. So it might be wise to have a discussion before jumping into implementation.

Reply:
`pthread_cleanup_push()/pthread_cleanup_pop()` are explicitly encouraged to allocate their data structure on the current thread stack though, which is a clever hack, but means that whatever mechanism is involved probably won't be sufficient for cleaning up the pthread data structures. Ideally, we'd have a callback that was explicitly *not* in the affected thread's context.

Reply:
> `pthread_cleanup_push()/pthread_cleanup_pop()` are explicitly encouraged to allocate their data structure on the current thread stack though, which is a clever hack, but means that whatever mechanism is involved probably won't be sufficient for cleaning up the pthread data structures. Ideally, we'd have a callback that was explicitly _not_ in the affected thread's context.

Ah yes... I forgot about that. In that case, I don't have an immediately better solution than what was done previously with `fn_abort`.

Reply:
So, we pretty clearly need to support a set of calls that happen upon thread termination. For the POSIX thread cleanup, and possibly freeing an allocated stack, we need calls that happen after thread termination and outside of the thread context. Do we also need to support calls that happen just before the thread is terminated that are made on the thread stack?

For the out-of-thread-context case, we can either have a per-thread callback list, or a global callback list. The former is "cleaner" in many ways; individual threads can have callbacks registered and those callbacks will be run with more constrained environments. However, the problem is where to allocate storage for the per-thread callback list. We could punt and just require that the caller provide suitable storage, or we could pre-allocate storage with the thread object itself -- otherwise the kernel would need to allocate somehow, and that seems bad.

With a global list, we would have a single list which could be allocated by the linker as an array. The callbacks would have to figure out if they were interested in the thread being terminated and somehow map that back to their related data structures, but that doesn't seem impractical to me.

Reply:
There are a lot of different ways we can try. If the desired solution is something that happens synchronously, in kernel context, then that's the kind of change that should really be made at the zephyr layer and not at the posix layer, and it should really be in its own PR.

I would consider this first PR as just satisfying the posix interface with an implementation that works with what is currently available. Implementation details can be changed / optimized later, so yeah, I would prefer to punt on synchronous callbacks.

Reply:
That's a good point -- we don't have any better mechanism than this today, and there's no reason to block this PR waiting for the perfect solution to appear.

So, what happens if the thread isn't completely finished within 100ms? Can we have the helper function block until the thread is actually finished so that we don't wait for the next `pthread_create` before things are cleaned up?

Reply:
That's a good question - technically we could create a barrier to be *very* sure it's serialized. I'm also quite happy making it a Kconfig and letting users adjust it to fit their needs, at least until a better kernel-level solution arises.

That reminds me - I should actually add the Kconfig option (which is what kicked off this thread in the first place) 🤣